lib/backend/k8s/conversion/conversion.go

```diff
@@ -468,14 +468,15 @@ func (c Converter) K8sNetworkPolicyToCalico(np *networkingv1.NetworkPolicy) (*mo
 	}
 
 	// Build and return the KVPair.
+	rev := c.JoinNetworkPolicyRevisions("", np.ResourceVersion)
```

Comment:
I'm not entirely convinced this is the right fix here - `JoinNetworkPolicyRevisions` is meant for resources that are backed by two separate APIs, where we need both revisions available.

In this case, NP is just a single API, and so this function isn't really doing anything. Especially without the corresponding "Split" revisions call. I would guess that it is coincidental that it stops the "Failed to create watcher" log mentioned in the original issue.

I think it was called out that the log is not an error, and expected periodically, so I'm not sure if there is anything to fix here except perhaps lowering the log level to debug?

Reply:
Let's discuss if this is an actual issue first. I'm not sure if I should post here, or in the bug. 

On our production system I see the `watchercache.go 259: Failed to create watcher` error about every minute or two. On our toy cluster, I will see this error message if and only if I update or create a k8s network policy. As you rightly point out, this is not immediate but periodic in some sense, I guess due to the previous watch expiring (I have not understood this bit in the code yet). The production system in question is essentially a build/test tool where namespaces and the associated network policies come and go, so seeing this error so often is not surprising.

The consequence is that it will perform a full resync very often. I know this because of the INFO log `2019-12-10 21:03:35.772 [INFO][8] networkpolicy.go 283: Returning NP KVPs KVPs=&model.KVPairList{KVPairs:[]*model.KVPair{(*model.KVPair)(0xc00ac2b280), … (*model.KVPair)(0xc009097b00)}, Revision:"1191809714/1191809718"}`. This log line contains the memory addresses for all NetworkPolicies in the cluster (about 850 for us right now).

So, as far as my investigation goes typha will perform a full resync on NetworkPolicies everytime any k8s NP is added or changed. Please point out if I misunderstand things here… because with my current level of knowledge it seems surprising if a full resync is desired behaviour here.

> Especially without the corresponding "Split" revisions call

To my understanding, the watchercache passes the "wrong" Revision (e.g. "1234") back to the client here: https://github.com/projectcalico/libcalico-go/blob/2d2a4ff81c856b51ee69705a26f72000252eb531/lib/backend/watchersyncer/watchercache.go#L238
I think the client is a `networkPolicyClient`, which calls into "Split" when setting up a watch: https://github.com/projectcalico/libcalico-go/blob/master/lib/backend/k8s/resources/networkpolicy.go#L322-L327
There's only two places that generate the `ResourceVersion is not valid` message `SplitNetworkPolicyRevision` and `SplitProfileRevision`, so it looks very likely to me that I'm hitting that code path.

> I would guess that it is coincidental that it stops the "Failed to create watcher" log mentioned in the original issue.
I think you're right. I looked through the code again at who calls `K8sNetworkPolicyToCalico`. There's three places in `networkPolicyClient`, one is in `Get`, one in `List` and one in `Watch`. The first two call `kvp.Revision = c.JoinNetworkPolicyRevisions("", kvp.Revision)` more or less immediately after. However, `Watch` doesn't: https://github.com/projectcalico/libcalico-go/blob/2d2a4ff81c856b51ee69705a26f72000252eb531/lib/backend/k8s/resources/networkpolicy.go#L343
Would a better fix be to also call `JoinNetworkPolicyRevisions` here?